### ISSUE TITLE
Synchronous XMLHttpRequest on the main thread is deprecated

### ISSUE THREAD

**Issue by odeleongt** [rstudio/shiny#789]
I'm getting the warning

```
Synchronous XMLHttpRequest on the main thread is deprecated because
of its detrimental effects to the end user's experience.
For more help, check http://xhr.spec.whatwg.org/.
```

when loading an app in the `shinyapps.io` service, due to the use of `c.async="false"` in column 16511 of `jquery.min.js` or `xml.async = "false";` in line 8567 of `jquery.js`.

It prevents the use of `$( document ).ready()` to execute js code when the document finishes loading.

**Comment by cperriard** [comment]
Hi odeleongt, 
do you use `selectInput(..., selectize = TRUE)` on the server side? 
I have the same problem with this minimal example.  
```r
library(shiny)
shinyApp(
  ui = shinyUI(fluidPage(
    uiOutput("in4ui")
  )), 
  server = function(input, output) {
    output$in4ui <- renderUI({
      selectInput('in4', 'Options', c(Choose='', LETTERS), selectize=TRUE)
    })
  }
)
```
I manually upgraded selectize.js and the warning remains. I think it must be some bug in one of the shiny javascript libraries. 

Using `selectInput(..., selectize = TRUE)` on the ui side results in no warning: 
```r
library(shiny)
shinyApp(
  ui = shinyUI(fluidPage(
    selectInput('in4', 'Options', c(Choose='', LETTERS), selectize=TRUE)
  )), 
  server = function(input, output) {
  }
)
```

**Comment by odeleongt** [comment]
It is not a bug. The problem is due to using outdated libraries. They need to ship the package with an updated version.

**Comment by jcheng5** [comment]
@wch Now that 0.12 is off to CRAN, maybe it's a good time to upgrade jQuery?

@odeleongt Are you sure this is true?

> It prevents the use of $( document ).ready() to execute js code when the document finishes loading.

I don't see any problem with or without the warning--and the deprecation warning doesn't mean any behavior has changed, they're just telling you to avoid it because they may eventually not support it, if I'm understanding correctly.

```r
library(shiny)
shinyApp(
  ui = shinyUI(fluidPage(
    uiOutput("in4ui"),
    tags$script(
      "$(document).ready(function(){alert('onready works');});"
    )
  )), 
  server = function(input, output) {
    output$in4ui <- renderUI({
      selectInput('in4', 'Options', c(Choose='', LETTERS), selectize=TRUE)
    })
  }
)
```

**Comment by odeleongt** [comment]
@jcheng5 Huh. It always worked locally, but failed on the shinyapps.io services. It does seem to work now.

**Comment by wch** [comment]
We could update jQuery, though I'd like to wait until after 0.12.0 is accepted by CRAN.

We're on jQuery 0.11.0, and the latest 1.x version is 0.11.3. @odeleongt Do you know if that version will make the message go away?

**Comment by odeleongt** [comment]
@wch Yep, it does. Either by recompiling the `shiny` package, or directly replacing the `jquery.js` and `jquery.min.js` files in the package directory using jQuery version 1.11.3..

**Comment by cperriard** [comment]
It is a matter how things are loaded. If you add a selectize input to the UI, the script tag for selectize.js is in the source and loaded by the browser directly. And the warning disappears. 
```r
library(shiny)

shinyApp(
  ui = shinyUI(fluidPage(
    uiOutput("in4ui"),
    selectInput('in5', 'Options UI', c(Choose='', LETTERS), selectize=TRUE)
  )
  ), 
  server = function(input, output) {
    output$in4ui <- renderUI({
      selectInput('in4', 'Options Server', c(Choose='', LETTERS), selectize=TRUE)
    })
  }
)
```
If I am right, scripts are loaded just by appending a script tag or HTML chunk to the head (srcjs/output_binding_html.js). jQuery `.append()` triggers the sync load of the script.  

If you run an empty shiny app, you could execute the following jQuery code to reproduce the warning:  
```javascript 
$( 'head' ).append( $( '<script src="shared/selectize/js/selectize.min.js"></script>' ) );
```
No warning, when loading css files: 
```javascript 
$( 'head' ).append( $( '<link rel="stylesheet" type="text/css" href="shared/selectize/css/selectize.bootstrap3.css">' ) );
```
```r
library(shiny)

shinyApp(
  ui = shinyUI(fluidPage(
  )
  ), 
  server = function(input, output) {
  }
)
```
Maybe it would be nice to load scripts async with a module loader like [require.js](http://requirejs.org/)

**Comment by pssguy** [comment]
I'm getting the same issue i.e app working fine locally but immediate greying out and the same deprecated message when run remotely [here](https://mytinyshinys.shinyapps.io/worldSoccer)

I have changed all the selectInputs (I think) to selectize=FALSE but that produces no change - other than looking less attractive

Not sure if it is linked, but as you can see if checking the console, lots of resources e.g. shared/shiny.css, jquery-ui.min.css etc. are not being loaded

Tested in Chrome and Firefox

**Comment by fereshtehRS** [comment]
@pssguy, your application on shinyapps.io reports some warnings/errors:

```
2015-08-04T17:22:30.759814646Z Warning in View(test) : unable to open display
a2015-08-04T17:22:30.760240380Z Warning: Unhandled error in observer: unable to start data viewer
)2015-08-04T17:22:30.760240380Z observe({
=2015-08-04T17:22:30.760240380Z     if (is.null(input$team)) 
02015-08-04T17:22:30.760240380Z         return()
S2015-08-04T17:22:30.760240380Z     selection <- all %>% filter(team == input$team)
`2015-08-04T17:22:30.760240380Z     selection <- cbind(selection, id = seq_len(nrow(selection)))
?2015-08-04T17:22:30.760240380Z     all_values <- function(x) {
82015-08-04T17:22:30.760240380Z         if (is.null(x)) 
82015-08-04T17:22:30.760240380Z             return(NULL)
c2015-08-04T17:22:30.760240380Z         row <- selection[selection$id == x$id, c("Season", "tier", 
82015-08-04T17:22:30.760240380Z             "Position")]
b2015-08-04T17:22:30.760240380Z         paste0(names(row), ": ", format(row), collapse = "<br />")
%2015-08-04T17:22:30.760240380Z     }
T2015-08-04T17:22:30.760240380Z     test <- selection %>% filter(team == input$team)
52015-08-04T17:22:30.760240380Z     print(View(test))
y2015-08-04T17:22:30.760240380Z     test %>% ggvis(~Season, ~Overall, `:=`(key, ~id)) %>% layer_points(fill = ~tier) %>% 
]2015-08-04T17:22:30.760240380Z         scale_numeric("y", reverse = TRUE) %>% add_axis("y", 
]2015-08-04T17:22:30.760240380Z         title = "Overall League Position") %>% add_axis("x", 
a2015-08-04T17:22:30.760240380Z         title = "", format = "####") %>% add_tooltip(all_values, 
k2015-08-04T17:22:30.760240380Z         "click") %>% handle_click(getSeason) %>% set_options(height = 480, 
K2015-08-04T17:22:30.760240380Z         width = 480) %>% bind_shiny("plot")
"2015-08-04T17:22:30.760240380Z })
M2015-08-04T17:22:30.760506049Z 2015/08/04 17:22:30 Error reading from Shiny:
```

See if you can find any issue in that piece of code that might not show itself locally.

**Comment by pssguy** [comment]
@fereshtehRS Thanks for that. I'll look into it, though it has been running fine locally
How did you access the above warnings?

**Comment by fereshtehRS** [comment]
@pssguy I got this on the server. But you might be able to see them using `shinyapps::showLogs()` in the IDE console.

**Comment by pssguy** [comment]
@fereshtehRS 
Yes my issue was actually a different one. A line which should not have been in my code but was not causing a local error. Thanks for your help. 
I still get the warning message but can use selectize=TRUE default with no detriment to the app

**Comment by fereshtehRS** [comment]
@pssguy Glad that you found the issue.

**Comment by kush99993s** [comment]
I have same issue, 
How do I have to solve this problem ?

There is solution in stackoverflow, http://stackoverflow.com/questions/28322636/synchronous-xmlhttprequest-warning-and-script, However, I don't know where to pass this argument.

**Comment by pvermees** [comment]
The `Synchronous XMLHttpRequest` error in my app can be traced back to RShiny's `getSettings()` function:
```
function getSettings() {
    var uri = window.location.pathname;
    if (uri.charAt(uri.length - 1) == '/')
        uri = uri.substr(0, uri.length - 1);
    var xhr = $.ajax({
        dataType: "json",
        url: uri + '/__settings__/',
        async: false
    });
    if (xhr.status == 200) {
        return JSON.parse(xhr.responseText)
    }
}
```

If synchronous calls are deprecated, then perhaps it would be good to remove the `async: false` bit in shinyapps.io and replace it with an asynchronous alternative? I use an HTML-only UI so the `selectize=TRUE` option mentioned above won't work for me.

**Comment by aronatkins** [comment]
Here is a simple Shiny application using `dateInput` that gives the XMLHttpRequest warning with Chrome.

```
library(shiny)
shinyApp(
  ui = shinyUI(fluidPage(
    uiOutput("in4ui")
  )), 
  server = function(input, output) {
    output$in4ui <- renderUI({
      dateInput('in4','date')
    })
  }
)
```

Adding the example here to show that it is not just `selectInput`.

**Comment by jcubic** [comment]
I also have this problem but the issue came from sockjs (I use XHR instead of sockets because we had problems with it).

> sockjs-0.3.4.min.js:27 [Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.

**Comment by mmuurr** [comment]
I was able to replicate this browser warning _specifically_ when using a custom input as part of a `showModal(modalDialog(...))` pattern. In my case, the input R code looks (roughly) like so, with the `singleton(...)` call being modeled after [this document](https://shiny.rstudio.com/articles/js-custom-input.html):
```R
my_input <- function(input_id) {
  withTags(tagList(
    singleton(head(script(src = "my-input.js"))),
    div(id = input_id, some, html, stuff, here)
  ))
}
```
* When `my_input` is part of the UI directly (e.g. via `ui <- fluidPage(my_input("foo"))`, there's no warning.
* When `my_input` is called via the modal sequence (e.g. via `showModal(modalDialog(my_input("foo")))`, however, the deprecation warning shows.

**Comment by jcheng5** [comment]
FYI, I investigate this every couple of years, I've never had any ideas for how to get past it. It comes from how jQuery's [append](https://api.jquery.com/append/) function deals with HTML that contains `<script>` tags, which we do [here](https://github.com/rstudio/shiny/blob/396dd2632e55d9dd48b98018cd272529d2a2382a/srcts/src/shiny/render.ts#L213); synchonous XHR is the only way jQuery can ensure that scripts load in the correct order.

[This jQuery issue](https://github.com/jquery/jquery/issues/1895) tracks the discussion, seems like changes might be coming in jQuery 4.0. To deal with this, we'd need to introduce asynchronous `renderValue`, which would cause async to ripple up the call stack, and we'd need to introduce some kind of queueing or mutex so we don't try to handle two server messages concurrently.

(Incidentally, not sure why but I'm no longer able to replicate this warning using Chrome 97, not even using test apps known to cause the warning using old versions of Shiny/jQuery.)

**Comment by jcubic** [comment]
Unfortunately, I can't help with reproduction, since we don't use XHR anymore. We had this problem when using a shiny server that required the use of XHR (I don't remember why). But we switched to our on infra based on Docker Swarm and the code use web sockets now.

If you want to see this error, I would just look at how the code sends Ajax requests, if it's through jQuery I would check if by any chance there is a sync call.

CHeck this SO question: [jQuery: Performing synchronous AJAX requests](https://stackoverflow.com/a/6685294/387194) I would look into places where this is happening. Sync requests are deprecated and it will always give warnings.

**Comment by jcubic** [comment]
I've used the new beta search [cs.github.com](https://cs.github.com/) and I can't find any `async: false` maybe the issue is with an older version of shiny.

@mmuurr which version of shiny you're using?

**Comment by wch** [comment]
This has been fixed by #3666.
